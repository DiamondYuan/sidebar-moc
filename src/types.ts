import vscode from "vscode";

export interface Point {
  line: number;
  column: number;
}

export type OutlineContent = OutlineHeadingNode | OutlineTreeNode;

export interface OutlineHeadingNode {
  type: "heading";
  startPoint?: Point;
  depth: number;
  text?: string;
  url?: string;
  children: Array<OutlineContent>;
}

export interface OutlineTreeNode {
  type: "treeNode";
  startPoint?: Point;
  text?: string;
  url?: string;
  children: Array<OutlineContent>;
}

export interface OutlineRoot {
  type: "root";
  children: Array<OutlineContent>;
}

export interface OutlineDataWithUri {
  uri: vscode.Uri;
  data: OutlineRoot | OutlineContent;
}
