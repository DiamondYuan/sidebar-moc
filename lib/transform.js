"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformMdastToMocAst = void 0;
var unist_util_select_1 = require("unist-util-select");
var mdast_util_to_string_1 = require("mdast-util-to-string");
function transformMdastToMocAst(root) {
    if (root.type === "root") {
        return {
            type: "root",
            children: transformChildren(root),
        };
    }
    else if (root.type === "list") {
        return transformChildren(root);
    }
    else if (root.type === "heading") {
        return __assign({ type: "heading", depth: root.depth, startPoint: getStartPoint(root.position), children: transformChildren(root) }, getHeadingContent(root));
    }
    else if (root.type === "listItem") {
        return __assign({ type: "treeNode", startPoint: getStartPoint(root.position), children: transformChildren(root) }, getListItemContent(root));
    }
    return null;
}
exports.transformMdastToMocAst = transformMdastToMocAst;
function getStartPoint(position) {
    if (!position) {
        return;
    }
    return {
        line: position.start.line,
        column: position.start.column,
    };
}
function transformChildren(root) {
    var children = [];
    var originalChildren = root.children;
    if (originalChildren.length > 0) {
        for (var _i = 0, originalChildren_1 = originalChildren; _i < originalChildren_1.length; _i++) {
            var child = originalChildren_1[_i];
            var result = transformMdastToMocAst(child);
            if (Array.isArray(result)) {
                children = children.concat(result);
            }
            else if (!!result) {
                children.push(result);
            }
        }
    }
    return children;
}
function getListItemContent(listItem) {
    var firstChild = null;
    for (var _i = 0, _a = listItem.children; _i < _a.length; _i++) {
        var item = _a[_i];
        if (item.type === "list" || item.type === "heading") {
            return;
        }
        if (!firstChild) {
            firstChild = item;
        }
        var linkItem = (0, unist_util_select_1.select)("link", item);
        if (linkItem) {
            return {
                text: (0, mdast_util_to_string_1.toString)(linkItem),
                url: linkItem.url,
            };
        }
    }
    if (firstChild) {
        return {
            text: (0, mdast_util_to_string_1.toString)(firstChild),
        };
    }
    return;
}
function getHeadingContent(heading) {
    var linkItem = (0, unist_util_select_1.select)("link", heading);
    if (linkItem) {
        return {
            text: (0, mdast_util_to_string_1.toString)(linkItem),
            url: linkItem.url,
        };
    }
    return {
        text: (0, mdast_util_to_string_1.toString)(heading),
    };
}
