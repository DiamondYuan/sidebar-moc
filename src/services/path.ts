import vscode from "vscode";

export interface IPathService {
  uriToConfigPath(fs: vscode.Uri): string;
  ConfigPathToUri(configPath: string): vscode.Uri;
}
