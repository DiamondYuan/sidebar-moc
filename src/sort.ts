import { OutlineRoot, OutlineContent, OutlineHeadingNode } from "./types";

export function sortRoot(root: OutlineRoot): OutlineRoot {
  return { type: "root", children: sortChildren(root.children) };
}

function sortChildren(nodes: OutlineContent[]): OutlineContent[] {
  const result: OutlineContent[] = [];
  let paths: number[] = [];
  for (const node of nodes) {
    let parentChildren = getParentChildren(result, paths);
    if (node.type === "treeNode") {
      parentChildren.children.push({
        ...node,
        children: sortChildren(node.children),
      });
    } else {
      while (
        parentChildren.parent !== null &&
        parentChildren.parent.depth >= node.depth
      ) {
        paths.pop();
        parentChildren = getParentChildren(result, paths);
      }
      parentChildren.children.push(node);
      paths.push(parentChildren.children.length - 1);
    }
  }
  return result;
}

function getParentChildren(
  root: OutlineContent[],
  path: number[]
): { children: OutlineContent[]; parent: OutlineHeadingNode | null } {
  let parent: OutlineHeadingNode | null = null;
  let children = root;
  if (path.length === 0) {
    return {
      children,
      parent,
    };
  }
  for (let item of path) {
    parent = children[item] as OutlineHeadingNode;
    children = parent.children;
  }
  return {
    children,
    parent,
  };
}
