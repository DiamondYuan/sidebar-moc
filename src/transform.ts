import { OutlineRoot, OutlineContent } from "./types";
import {
  Content as MdastContent,
  Root as MdastRoot,
  Heading,
  List,
  ListItem,
} from "mdast";

export function transformMdastToMocAst(
  root: MdastContent | MdastRoot
): OutlineRoot | null | OutlineContent | OutlineContent[] {
  if (root.type === "root") {
    return {
      type: "root",
      children: transformChildren(root),
    };
  } else if (root.type === "list") {
    return transformChildren(root);
  } else if (root.type === "heading") {
    return {
      type: "heading",
      depth: root.depth,
      startPoint: getStartPoint(root.position),
      children: transformChildren(root) as any,
    };
  } else if (root.type === "listItem") {
    return {
      type: "treeNode",
      startPoint: getStartPoint(root.position),
      children: transformChildren(root) as any,
    };
  }
  return null;
}

function getStartPoint(position?: Heading["position"]) {
  if (!position) {
    return;
  }
  return {
    line: position.start.line,
    column: position.start.column,
  };
}

function transformChildren(
  root: Heading | List | MdastRoot | ListItem
): OutlineContent[] {
  let children: any[] = [];
  let originalChildren = root.children;
  if (originalChildren.length > 0) {
    for (const child of originalChildren) {
      const result = transformMdastToMocAst(child);
      if (Array.isArray(result)) {
        children = children.concat(result);
      } else if (!!result) {
        children.push(result);
      }
    }
  }
  return children;
}
