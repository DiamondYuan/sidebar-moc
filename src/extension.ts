import { TOCCompletionItemProvider } from "./contrib/language/path-completion-item-provider";
import { TocService } from "./services/toc-service";
import { Point, OutlineDataWithUri } from "./types";
import vscode, { Uri } from "vscode";
import { TOCTreeDataProvider } from "./contrib/toc-tree-data-provider";
import { Utils } from "vscode-uri";

function updateContext() {
  const config = vscode.workspace.getConfiguration();
  const mocPaths: string[] = config.get("sidebar-moc.mocPath") ?? [];
  const fsPath = vscode.window.activeTextEditor?.document?.uri.fsPath;
  if (!fsPath) {
    return;
  }
  if (mocPaths.includes(fsPath)) {
    vscode.commands.executeCommand("setContext", "inSideBarMoc", true);
  } else {
    vscode.commands.executeCommand("setContext", "inSideBarMoc", false);
  }
}

export async function activate(context: vscode.ExtensionContext) {
  const tocService = new TocService();
  const treeDataProvider = new TOCTreeDataProvider(tocService);
  tocService.onTocChange(() => {
    treeDataProvider.refresh();
  });
  vscode.window.createTreeView("sidebar-moc.view", {
    treeDataProvider: treeDataProvider,
  });
  const watcher = vscode.workspace.createFileSystemWatcher("**/*.md");
  watcher.onDidChange((e) => {
    tocService.updateToc(e);
  });
  var selector: vscode.DocumentSelector = [
    {
      pattern: "**",
    },
  ];
  context.subscriptions.push(
    vscode.languages.registerCompletionItemProvider(
      selector,
      new TOCCompletionItemProvider(tocService),
      "/"
    )
  );
  vscode.commands.registerCommand(
    "sidebar-moc.open-uri",
    (file: { uri: string; outlineData: OutlineDataWithUri["data"] }) => {
      const mocUri = Uri.file(file.uri);
      if (file.outlineData.type === "root") {
        openAndShow(mocUri, { line: 1, column: 1 });
        return;
      }
      if (!file.outlineData.url) {
        openAndShow(mocUri, file.outlineData.startPoint);
        return;
      }
      if (file.outlineData.url.startsWith("http")) {
        vscode.env.openExternal(vscode.Uri.parse(file.outlineData.url));
        return;
      }
      const fileUrI = vscode.Uri.joinPath(
        Utils.dirname(mocUri),
        file.outlineData.url
      );
      openAndShow(fileUrI);
    }
  );

  vscode.window.onDidChangeActiveTextEditor((e) => {
    updateContext();
  });

  vscode.workspace.onDidChangeConfiguration(() => {
    const config = vscode.workspace.getConfiguration();
    let paths: string[] = config.get("sidebar-moc.mocPath") ?? [];
    tocService.updateConfig(paths);
  });

  vscode.commands.registerCommand("sidebar-moc.add-moc", (e: Uri) => {
    const config = vscode.workspace.getConfiguration();
    const paths: string[] = config.get("sidebar-moc.mocPath") ?? [];
    if (!paths.includes(e.fsPath)) {
      paths.push(e.fsPath);
    }
    config.update("sidebar-moc.mocPath", paths).then(() => {
      updateContext();
    });
  });
  vscode.commands.registerCommand("sidebar-moc.remove-moc", (e: Uri) => {
    const config = vscode.workspace.getConfiguration();
    let paths: string[] = config.get("sidebar-moc.mocPath") ?? [];
    if (paths.includes(e.fsPath)) {
      paths = paths.filter((o) => o !== e.fsPath);
    }
    config.update("sidebar-moc.mocPath", paths).then(() => {
      updateContext();
    });
  });
}

function openAndShow(file: vscode.Uri, point?: Point) {
  vscode.workspace.openTextDocument(file).then(
    (a: vscode.TextDocument) => {
      if (point) {
        vscode.window.showTextDocument(a, {
          selection: point
            ? new vscode.Range(
                new vscode.Position(point.line - 1, point.column - 1),
                new vscode.Position(point.line - 1, point.column - 1)
              )
            : undefined,
        });
      } else {
        vscode.window.showTextDocument(a);
      }
    },
    (error: any) => {
      //
    }
  );
}
