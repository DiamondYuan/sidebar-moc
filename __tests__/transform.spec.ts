import { transformMdastToMocAst } from "../src/transform";
import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkGfm from "remark-gfm";
import path from "path";
import fs from "fs";

unified().use(remarkParse).use(remarkGfm).parse("");

const summary = path.join(__dirname, "fixture", "summary.md");
const summaryContent = fs.readFileSync(summary, "utf-8");

it("", () => {
  const mdast = unified().use(remarkParse).use(remarkGfm).parse(summaryContent);
  expect(transformMdastToMocAst(mdast)).toMatchSnapshot();
});
