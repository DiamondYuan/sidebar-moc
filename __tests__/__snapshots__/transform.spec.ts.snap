// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "depth": 1,
      "startPoint": Object {
        "column": 1,
        "line": 1,
      },
      "type": "heading",
    },
    Object {
      "children": Array [],
      "depth": 3,
      "startPoint": Object {
        "column": 1,
        "line": 3,
      },
      "type": "heading",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "startPoint": Object {
            "column": 3,
            "line": 6,
          },
          "text": "Writing is nice",
          "type": "treeNode",
          "url": "part1/README.md#writing",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "startPoint": Object {
                "column": 5,
                "line": 8,
              },
              "text": "GitBook is nice",
              "type": "treeNode",
              "url": "part1/README.md#gitbook",
            },
          ],
          "startPoint": Object {
            "column": 3,
            "line": 7,
          },
          "text": "GitBook is nice",
          "type": "treeNode",
          "url": "part1/README.md#gitbook",
        },
      ],
      "startPoint": Object {
        "column": 1,
        "line": 5,
      },
      "text": "Part I",
      "type": "treeNode",
      "url": "part1/README.md",
    },
  ],
  "type": "root",
}
`;
